clusterloader2/pkg/test/simple_test_executor.go

```diff
@@ -128,83 +128,88 @@ func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) *util
-		var maxCurrentReplicaCount int32
-		for j := range instancesStates {
-			if instancesStates[j].CurrentReplicaCount > maxCurrentReplicaCount {
-				maxCurrentReplicaCount = instancesStates[j].CurrentReplicaCount
-			}
-		}
-		// Deleting objects with index greater or equal requested replicas per namespace number.
-		// Objects will be deleted in reversed order.
-		for replicaIndex := phase.ReplicasPerNamespace; replicaIndex < maxCurrentReplicaCount; replicaIndex++ {
-			for j := len(phase.ObjectBundle) - 1; j >= 0; j-- {
-				if replicaIndex < instancesStates[j].CurrentReplicaCount {
-					<-ticker.C
-					if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex, DELETE_OBJECT); !objectErrList.IsEmpty() {
-						errList.Concat(objectErrList)
-						if isErrsCritical(objectErrList) {
-							return errList
-						}
-					}
+			// Calculating maximal replica count of objects from object bundle.
+			var maxCurrentReplicaCount int32
+			for j := range instancesStates {
+				if instancesStates[j].CurrentReplicaCount > maxCurrentReplicaCount {
+					maxCurrentReplicaCount = instancesStates[j].CurrentReplicaCount
 				}
 			}
-		}
-		// Handling for update/create objects.
-		for replicaIndex := int32(0); replicaIndex < phase.ReplicasPerNamespace; replicaIndex++ {
-			for j := range phase.ObjectBundle {
-				if instancesStates[j].CurrentReplicaCount == phase.ReplicasPerNamespace {
-					<-ticker.C
-					if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex, PATCH_OBJECT); !objectErrList.IsEmpty() {
-						errList.Concat(objectErrList)
-						if isErrsCritical(objectErrList) {
-							return errList
+
+			var namespaceWG wait.Group
+			// Deleting objects with index greater or equal requested replicas per namespace number.
+			// Objects will be deleted in reversed order.
+			for replicaCounter := phase.ReplicasPerNamespace; replicaCounter < maxCurrentReplicaCount; replicaCounter++ {
+				replicaIndex := replicaCounter
+				namespaceWG.Start(func() {
+					for j := len(phase.ObjectBundle) - 1; j >= 0; j-- {
+						if replicaIndex < instancesStates[j].CurrentReplicaCount {
+							<-ticker.C
+							if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex, DELETE_OBJECT); !objectErrList.IsEmpty() {
+								errList.Concat(objectErrList)
+							}
 						}
-						// If error then skip this bundle
-						break
 					}
-				} else if replicaIndex >= instancesStates[j].CurrentReplicaCount {
-					<-ticker.C
-					if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex, CREATE_OBJECT); !objectErrList.IsEmpty() {
-						errList.Concat(objectErrList)
-						if isErrsCritical(objectErrList) {
-							return errList
+				})
+			}
+			// Handling for update/create objects.
+			for replicaCounter := int32(0); replicaCounter < phase.ReplicasPerNamespace; replicaCounter++ {
```

Comment:
It seems like you are introducing goroutines proportional to #namespaces x #replicas. This can be a really huge number of them and I'm not sure if you should have such uncontrolled parallelism. Instead, bound parallelism / fixed no. of workers is better IMO.

@wojtek-t - for thoughts

Reply:
go-routines is fairly cheap mechanism - it's completely fine to have thousands of them running at the same time.
Didn't look into the code, but assuming most of them are sleeping at any given point in time, that sounds fine to me.

Reply:
You're right. Since most of them would be blocked on timer channel at all times, that should be ok.

Couple of points though:

- I'm not sure if it makes sense to have more goroutines than our qps (which will be a bottleneck anyway)
- There is some small constant [memory overhead associated](https://stackoverflow.com/questions/8509152/max-number-of-goroutines) with each goroutine - about 4MB per 1000 goroutines - but we may be able to live with that